gcc/config/riscv/riscv.cc

```diff
@@ -4593,18 +4613,45 @@ riscv_emit_stack_tie (void)
     emit_insn (gen_stack_tiedi (stack_pointer_rtx, hard_frame_pointer_rtx));
 }
 
+static void
+riscv_adjust_vector_frame (rtx target, poly_int64 offset)
```

Comment:
Hi @zhongjuzhe  ,
  I have some questions about the frame_info and stack  of variable length type for your help. Could you give me some advice on this?
In the current gcc framework, it seems that 'poly_uint64(coeff0, coeff1)' is used to calculate 'cfun->machine->frame', but only one variable length type(vector) can be processed.
  If I want to add a new variable length type, how can I handle its frame_info and stack? In other words, how to handle the frame_info and stack of various variable length MACHINE_MODE.
  Thanks, looking forward to your comments.

Reply:
I am not sure I am on the same page with you.
Are you asking how to handle the randomly various size of stack that includes different size of scalable vector.
If so, the answer is:
the  "poly_int64 offset" is the total size of the stack we need to assign.
For example, in "rvv-next" (not the same as upstream GCC that I support), 
"vint32m1_t" vector type has the size of [16,16], "vint32mf2_t" has vector type has the size of "[8,8].

If you want to assign a stack for for example: 2 "vint32m1_t" and 3 "vint32mf2_t" and additionl what  ever bytes (let's say 17 bytes). Here the total size of stack = 2 *[16,16] + 3 * [8,8] + 17 bytes. Then the poly_int64 offset = [73,56].  
riscv_adjust_vector_frame will handle this size of stack using csrr vlenb + scalar instructions to calculate the offset, and then assign this stack by sp - offset. Now riscv_adjust_vector_frame can handle any combination of poly_int64 value.

If you want to add a new variable length MACHINE_MODE that this rvv-next doesn't have, For example: VNx2TImode.
You just need to add it and use ADJUST_BYTES it correctly.  For example, suppose the VNx2TI occupy 2 RVV vector registers. You can just add this mode in riscv-modes.def:
VECTOR_MODE (VNx2TI)
ADJUST_BYTES (VNx2TI, poly_uint16 (16 * 2, 16 * 2) ([16,16]denotes a single vector register)

Then no more work need to be done.

Reply:
Thank you for your quick and detailed reply, which has greatly benefited me.
  I understand the function riscv_adjust_vector_frame according to your reply, but there is still a question. Can you help me again?
  I want to add a new variable length "matrix" register, similar to the vector register, which's MACHINE_MODE  is VECTOR_MODE_P. Similar to "csrr vlenb", "csrr mlenb" is used. The mlenb represents the length of the matrix register.  Everything goes smoothly. But when matrix and vector are used together, it is difficult to distinguish when to use "csrr vlenb" and when to use "csrr mlenb" in  the function riscv_adjust_vector_frame, mainly because there is no relationship between mlenb and vlenb, and the poly_int64 offset seems unable to represent more than one variable length type. 
 I tried several times, but still couldn't find a good solution. Do I have a chance to get your comments again? Or if you have any doubts about my question, I can add them.
  Thanks.

Reply:
OK. I understand your want to support your private "Matrix" extension in RVV GCC.

1. To design your own "Matrix" extension base on RVV. I think you should study and reference to"SME" (ARM Matrix extension). It's easily obtained in google search. Their ISA design is nice and clean. I would suggest you study it and redesign "Matrix" extension in RISC-V following "SME". But I could tell your basic ideas about "SME":
    -  SME is base SVE.  Same, your can design RISC-V Matrix extension (I call it RVM may be). RVM base RVV.
    -  When SME is enabled, SVE is enabled too by default. Same to RVM and RVV.
    -  Matrix register size in SME is strictly related to vector length of SVE. The Matrix register size = row size * column 
       size. 
    -  The relationship between SME matrix register size and SVE vector register size:
        row size of "Matrix" register = column size of "Matrix" register = Machine vector length of SVE vector.
        So by copying the idea of SME, I suggest you define your Matrix register size = vlenb * vlenb in RVV.
   -   Base on the assumption I listed above, there is no need to introduce a new instruction "csrr mlenb". Because 
       "csrr vlenb" can also handle "Matrix" register variable length. After you studied ISA of SME, you will learned my
       idea. There is no new instructions of handling variable length introduced in SME. Actually, all instructions in SME
       are calculation operations (for example. like matmul). In SME, they reuse the variable length handling instructions 
       belong SVE that's why they need to enable SVE by default for SME. 

     So I highly recommand you redesign your Matrix extension following SME. 
     It's very simple, Let's say define matrix register size = vlenb * vlenb. Then add a instruction like matmul. 
     Define a matrix machine mode in riscv-modes.def: VECTOR_MODE (Matrix mode).
     Then adjust mode size: ADJUST_BYTES (VLENB * VLENB, VLENB * VLENB).
     To handle this matrix register, you don't need distinguish them with RVV. They can be put together be handled.
      The current RVV GCC can well handle that for example assign a stack for  2 Matrix register + 3 vector register + 13 
      bytes. To handle that, the instructions sequence will be : 
      csrr a5, vlenb
      mult a6, a5, a5
      mult a5, a5, 3
      add  a5, a6, a5
      addi a5, a5, 13
      Then the value of a5 is the total stack size that you need to assign. Also to speed up this sequence, you could 
      introduce "csrr a5,mlenb" and revise poly handling in RVV GCC,  
     "csrr a5,mlenb" shoule be equal to "csrr a5,vlenb + mult a5,a5,a5".
     

2.  There is no mature solution to handle two or more varible length in compile, no matter LLVM or GCC. 
    Because so far no one need to do it.

4. If you really want to handle 2 or more variable length (Or I prefer calling it handle 2 or more runtime invariant), 
    I have no idea. It will make compile so complicated and there is no one has ever tried in both LLVM and GCC. 
    Maybe "#define NUM_POLY_INT_COEFFS 3" define it as 3? I am not sure.  
    To learn more about runtime invariant in GCC, I suggest you read this:
    https://gcc.gnu.org/onlinedocs/gccint/poly_005fint.html#poly_005fint

Reply:
Thank you for your suggestion. 

Since I want the matrix extension to be completely decoupled from the vector extension, I do not plan to follow the SME approach. 

I am using "#define NUM_POLY_INT_COEFFS 3" to achieve my goal. It is basically OK now.  As you said, I need to refactor the poly handling, but since poly may have taken into account the subsequent expansion when it was designed, the workload does not seem to be as large as expected.

Thanks again :)